tests/Tests/Unit/SerializerTest.php

```diff
@@ -291,4 +291,50 @@ public function testSpecialEncodingDecodingByFieldType()
         $this->assertEqualsWithDelta($data['blue'], $array['blue'], 0.0000001);
         $this->assertEqualsWithDelta($data['alpha']['value'], $array['alpha']['value'], 0.0000001);
     }
+
+    /**
+     * @dataProvider customEncoderProvider
+     */
+    public function testCustomEncoderForEncodeMessage($customEncoder, $protoObj, $expectedData)
+    {
+        $serializer = new Serializer([], [], [], [], $customEncoder);
+
+        $data = $serializer->encodeMessage($protoObj);
+        array_walk($expectedData, function ($val, $key, $expectedData) {
+            $this->assertEquals($val, $expectedData[$key]);
+        }, $expectedData);
+    }
+
+    public function customEncoderProvider()
+    {
+        $original = [
+            'red' => 50.0,
+            'blue' => 50.0
+        ];
+
+        $modified = [
+            'red' => 100,
+            'blue' => 100
+        ];
+
+        return  [
+            [
+                [
+                    Color::class => function ($message) use($modified) {
+                        return $modified;
+                    }
```

Comment:
nit, but can we make this serializer do something with the message and fields? This example is a little too contrived for me. Something like this:

```php
Color::class => function (Color $message) {
    return ['red' => $message->getRed() * 2, 'blue' => $message->getBlue() * 2];
}
```